pkg/trojan/message.go

```diff
@@ -0,0 +1,283 @@
+// the chunk address will have one of the targets as its prefix and thus will be forwarded to the neighbourhood of the recipient overlay address the target is derived from
+func (m *Message) Wrap(targets Targets) (swarm.Chunk, error) {
+	if err := checkTargets(targets); err != nil {
+		return nil, err
+	}
+
+	span := make([]byte, 8)
+	// 4064 bytes for trojan message + 32 bytes for nonce = 4096 bytes as payload for resulting chunk
+	binary.LittleEndian.PutUint64(span, swarm.ChunkSize)
+
+	return m.toChunk(targets, span)
+}
+
+// Unwrap creates a new trojan message from the given chunk payload
+// this function assumes the chunk has been validated as a content-addressed chunk
+// it will return the resulting message if the unwrapping is successful, and an error otherwise
+func Unwrap(c swarm.Chunk) (*Message, error) {
+	d := c.Data()
+
+	// unmarshal chunk payload into message
+	m := new(Message)
+	// first 40 bytes are span + nonce
+	err := m.UnmarshalBinary(d[40:])
+
+	return m, err
+}
+
+// IsPotential returns true if the given chunk is a potential trojan
+func IsPotential(c swarm.Chunk) bool {
+	// chunk must be content-addressed to be trojan
+	if c.Type() != swarm.ContentAddressed {
+		return false
+	}
+
+	data := c.Data()
+	// check for minimum chunk data length
+	// span (8) + nonce (32) + length (2) + topic (32) = 74
+	if len(data) < 74 {
+		return false
+	}
+
+	// check for valid trojan message length in bytes #41 and #42
+	messageLen := int(binary.BigEndian.Uint16(data[40:42]))
+	return 74+messageLen <= len(data)
+}
+
+// checkTargets verifies that the list of given targets is non empty and with elements of matching size
+func checkTargets(targets Targets) error {
+	if len(targets) == 0 {
+		return ErrEmptyTargets
+	}
+	validLen := len(targets[0]) // take first element as allowed length
+	for i := 1; i < len(targets); i++ {
+		if len(targets[i]) != validLen {
+			return ErrVarLenTargets
+		}
+	}
+	return nil
+}
+
+// toChunk finds a nonce so that when the given trojan chunk fields are hashed, the result will fall in the neighbourhood of one of the given targets
+// this is done by iteratively enumerating different nonces until the BMT hash of the serialization of the trojan chunk fields results in a chunk address that has one of the targets as its prefix
+// the function returns a new chunk, with the found matching hash to be used as its address,
+// and its data set to the serialization of the trojan chunk fields which correctly hash into the matching address
+func (m *Message) toChunk(targets Targets, span []byte) (swarm.Chunk, error) {
+	// start out with random nonce
+	nonce := make([]byte, 32)
+	if _, err := rand.Read(nonce); err != nil {
+		return nil, err
+	}
+	nonceInt := new(big.Int).SetBytes(nonce)
+	targetsLen := len(targets[0])
+
+	// serialize message
+	b, err := m.MarshalBinary() // TODO: this should be encrypted
```

Comment:
why should it be encrypted? isn't that up to whoever is the client of this code?

Reply:
i think [canonically](https://github.com/ethersphere/swarm/issues/2160), encryption is part of the whole trojan business.

having said that, this would be a good discussion to have. i guess clients could make this choice for themselves; whether to use encryption or not

Reply:
That is a `pss` construct then, not a `trojan` construct.

Reply:
@nolash @mortelli TBH trojan message itself  is a `pss` construct

Reply:
@zelig , agreed. 
`/pkg/pss/trojan/trojan.go`

Reply:
i would say leave as is until we tackle pss

Reply:
> @zelig , agreed.
/pkg/pss/trojan/trojan.go

I was hoping for the other answer; that `trojan` and `pss` themselves could be kept apart as well-defined concepts.

Reply:
i agree with @nolash on this one

Reply:
Please, don't take my technical code comments as leading. Depending on you guys here :).

Reply:
@jmozah the todo comment is still there...